### ISSUE TITLE
Error on MacOS related to custom class registration in torchscript

### ISSUE THREAD

**Issue by prabhat00155** [pytorch/pytorch#65000]
## 🐛 Bug

I am trying to build torchvision with latest ffmpeg(4.3). The build goes through, but when I import torchvision, it crashes with the following:
```
>>> import torchvision
libc++abi: terminating with uncaught exception of type c10::Error: Type c10::tagged_capsule<vision::video::Video> could not be converted to any of the known types.
Exception raised from operator() at /Users/prabhatroy/miniconda/envs/ffmpeg_py38/lib/python3.8/site-packages/torch/include/ATen/core/jit_type.h:1673 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 98 (0x1165d5c82 in libc10.dylib)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 106 (0x1165d439a in libc10.dylib)
frame #2: c10::detail::getTypePtr_<c10::tagged_capsule<vision::video::Video> >::call()::'lambda'()::operator()() const + 263 (0x1402f93f7 in video_reader.so)
frame #3: c10::detail::getTypePtr_<c10::tagged_capsule<vision::video::Video> >::call() + 27 (0x1402f925b in video_reader.so)
frame #4: c10::detail::infer_schema::(anonymous namespace)::createArgumentVector(c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>) + 216 (0x11bfef5b8 in libtorch_cpu.dylib)
frame #5: c10::detail::infer_schema::make_function_schema(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&&, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>, c10::ArrayRef<c10::detail::infer_schema::ArgumentDef>) + 123 (0x11bfef35b in libtorch_cpu.dylib)
frame #6: torch::jit::Function* torch::class_<vision::video::Video>::defineMethod<torch::class_<vision::video::Video>& torch::class_<vision::video::Video>::def<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(torch::detail::types<void, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::initializer_list<torch::arg>)::'lambda'(c10::tagged_capsule<vision::video::Video>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::initializer_list<torch::arg>) + 493 (0x1402f84dd in video_reader.so)
frame #7: torch::class_<vision::video::Video>& torch::class_<vision::video::Video>::def<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(torch::detail::types<void, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::initializer_list<torch::arg>) + 99 (0x1402e55d3 in video_reader.so)
frame #8: _GLOBAL__sub_I_video.cpp + 174 (0x14030bdae in video_reader.so)
frame #9: 0x0 + 4692351815 (0x117af9b47 in ???)
frame #10: 0x0 + 4692352850 (0x117af9f52 in ???)
frame #11: 0x0 + 4692331238 (0x117af4ae6 in ???)
frame #12: 0x0 + 4692322463 (0x117af289f in ???)
frame #13: 0x0 + 4692322624 (0x117af2940 in ???)
frame #14: 0x0 + 4692257298 (0x117ae2a12 in ???)
frame #15: 0x0 + 4692304154 (0x117aee11a in ???)
frame #16: dlopen_internal(char const*, int, void*) + 185 (0x7fff20423c94 in libdyld.dylib)
frame #17: dlopen + 28 (0x7fff2041207e in libdyld.dylib)
frame #18: py_dl_open + 153 (0x10ef0cbd9 in _ctypes.cpython-38-darwin.so)
<omitting python frames>

Abort trap: 6
```

This crash only happens on MacOS. If I remove the [registration of custom class](https://github.com/pytorch/vision/blob/bc3f8f6c134e545edb06dd830bba0683797d1a66/torchvision/csrc/io/video/video.cpp#L333), I don't see the crash.

## To Reproduce

```
conda create --name vision_env_py38 python=3.8
conda activate vision_env_py38
conda install -c conda-forge ffmpeg
conda install pytorch -c pytorch-nightly
git clone https://github.com/pytorch/vision.git
cd vision
git apply setup_patch.txt
cd ..
python -c "import torchvision"
```
[setup_patch.txt](https://github.com/pytorch/pytorch/files/7187224/setup_patch.txt)

## Expected behavior

Since the custom class is being registered [here](https://github.com/pytorch/vision/blob/bc3f8f6c134e545edb06dd830bba0683797d1a66/torchvision/csrc/io/video/video.cpp#L333), I would not expect this error:
```
libc++abi: terminating with uncaught exception of type c10::Error: Type c10::tagged_capsule<vision::video::Video> could not be converted to any of the known types.
```

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
Collecting environment information...
PyTorch version: 1.10.0.dev20210914
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.5.1 (x86_64)
GCC version: Could not collect
Clang version: 12.0.5 (clang-1205.0.22.11)
CMake version: version 3.19.6
Libc version: N/A

Python version: 3.8.11 (default, Aug  6 2021, 08:56:27)  [Clang 10.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.10.0.dev20210914
[pip3] torchvision==0.11.0a0+719e120
[conda] blas                      1.0                         mkl  
[conda] mkl                       2021.3.0           hecd8cb5_517  
[conda] numpy                     1.21.2                   pypi_0    pypi
[conda] pytorch                   1.10.0.dev20210914         py3.8_0    pytorch-nightly
[conda] torchvision               0.11.0a0+719e120          pypi_0    pypi
```

```
conda list | grep ffmpeg
ffmpeg                    4.3.2                h4dad6da_0    conda-forge
```

**Comment by mthrok** [comment]
Does this happen on other versions of Python?

On my M1 mac, I could not reproduce this with Python 3.8 and 3.9.
But on my Intel mac, I could reproduce with Python 3.8. I tried to reproduce with other Python versions but I got an NumPy dependency issue so I could not.

I am curious to see what will happen if you replace the implementation of `Video` class with some dummy class. We should find out if this is caused by PyTorch core or ffmpeg specific to Python 3.6 on macOS.

**Comment by prabhat00155** [comment]
I could reproduce this issue on Python version 3.6, 3.7, 3.8 and 3.9 on my Mac.
I saw the numpy dependency error with Python 3.6, but installing numpy fixes that.

**Comment by andfoy** [comment]
I did some debugging related to this issue under the following scenarios, all they imply the compilation of torchvision from source

1. Installing PyTorch from the nightly channel and FFmpeg from the PyTorch channel: Error reproduced
2. Compiling PyTorch in debug mode and compiling FFmpeg in debug/release mode: No error
3. Compiling PyTorch in debug mode and installing FFmpeg from the PyTorch channel: No error
4. Installing PyTorch from the nightly channel and compiling FFmpeg in debug/release mode: Error reproduced

From this it is possible to infer that there is some kind of incompatibility between the `video_reader` extension binaries and the nightly PyTorch distribution.

**Comment by prabhat00155** [comment]
I also see this issue even when building torchvision without ffmpeg, if system ffmpeg(`brew install ffmpeg`) is present.

**Comment by Luthaf** [comment]
I have the same issue, and it seems to be related to the registration of the constructor. The library loads fine with this code:

```cpp
class Example: public torch::CustomClassHolder  {
public:
    Example(std::string, std::string) {}
};

TORCH_LIBRARY(example, m) {
    m.class_<Example>("Example");
}
```

but if I define a constructor I get the same error message `Type c10::tagged_capsule<Example> could not be converted to any of the known types`.

```cpp
TORCH_LIBRARY(example, m) {
    m.class_<Example>("Example")
        .def(torch::init<std::string, std::string>());
}
```

---

This code works fine with torch 1.9, but fails with torch 1.10 or the latest nightly (torch-1.11.0.dev20211109).

**Comment by Luthaf** [comment]
The issue is fixed for me with the latest nightly (torch-1.11.0.dev20220107)

**Comment by bjuncek** [comment]
Fixed with latest release so closing the issue
